Add SQL reference pages for physical cluster replication (#18041)
cockroachdb/docs: src/current/v23.2/show-virtual-cluster.md

---
title: SHOW VIRTUAL CLUSTER

> summary: The SHOW VIRTUAL CLUSTER statement provides details on virtual clusters and physical replication streams.

Author: Yeah, this is funny, this has definitely come up as feedback before... not sure if that is a UK English thing (or a weird me thing) coming in there. Will keep this in mind. Updated!

toc: true
docs_area: reference.sql
---

{{site.data.alerts.callout_info}}
{% include feature-phases/preview.md %}
{{site.data.alerts.end}}

> {% include_cached new-in.html version="v23.2" %} The `SHOW VIRTUAL CLUSTER` statement lists the virtual clusters running in a CockroachDB cluster. `SHOW VIRTUAL CLUSTER` only supports inspecting virtual cluster status as part of the [physical cluster replication]({% link {{ page.version.version }}/physical-cluster-replication-overview.md %}) workflow.

Reviewer: Is it worth adding this text to an include and using that on all the pages? I can show you how to do a parameterized include if you want it to say slightly different things on different pages.

Author: I'm not sure, for the same reason as a few comments above, I kind of think untangling this at a later point might be annoying.

## Required privileges

`SHOW VIRTUAL CLUSTER` requires the `admin` role or the `MANAGEVIRTUALCLUSTER` [system privilege]({% link {{ page.version.version }}/security-reference/authorization.md %}#privileges).

Use the [`GRANT SYSTEM`]({% link {{ page.version.version }}/grant.md %}) statement:

{% include_cached copy-clipboard.html %}
~~~ sql
GRANT SYSTEM MANAGEVIRTUALCLUSTER TO user;
~~~

{% include enterprise-feature.md %}

## Synopsis

<div>
{% include {{ page.version.version }}/physical-replication/show-virtual-cluster-diagram.html %}
</div>

## Parameters

Parameter | Description
----------+------------
`virtual_cluster_spec` | The name of the virtual cluster.

> `REPLICATION STATUS` | Display the details of a replication stream.

Reviewer: So in a single operation, you can show the status or the capabilities, but not both at once? (there is no `ALL` type shortcut?)

Author: Actually, thanks for mentioning this, you can do both as it turns out (just tested), so `SHOW VIRTUAL CLUSTER(S) WITH REPLICATION STATUS, CAPABILITIES`. I edited the diagram for this

`CAPABILITIES` | Display the [capabilities]({% link {{ page.version.version }}/create-virtual-cluster.md %}#capabilities) of a virtual cluster.

{% comment %}Responses and Data state are a copy from the monitoring page, these should be in an include once the first PR has merged.{% endcomment %}
## Responses

> The following table is a list of all the responses returned with the different `SHOW VIRTUAL CLUSTER` statements:

Reviewer: ```suggestion
This table lists all possible responses from the different `SHOW VIRTUAL CLUSTER` statements:
```

Author: Changed!

Field    | Response
---------+----------
`id` | The ID of the virtual cluster.
`name` | The name of the standby (destination) virtual cluster.
`data_state` | The state of the data on the virtual cluster. This can show one of the following: `initializing replication`, `ready`, `replicating`, `replication paused`, `replication pending cutover`, `replication cutting over`, `replication error`. Refer to [Data state](#data-state) for more detail on each response.

> `service_mode` | The service mode shows whether the virtual cluster is ready to accept SQL requests. This can show one of the following: <br>`none`<br>`shared`: The virtual cluster's SQL connections will be served by the same nodes that are serving the system interface.

Reviewer: I think this line has some formatting issues .. is the `:` a stray? I think you probably wanted another `<br>` there.

Author: I adapted this. It actually wasn't a formatting issue, but realized it wasn't working out how I wanted it to.

`source_tenant_name` | The name of the primary (source) virtual cluster.

> `source_cluster_uri` | The URI of the primary (source) cluster. This is the URI that connects to the primary cluster to [start a replication stream]({% link {{ page.version.version }}/set-up-physical-cluster-replication.md %}#step-4-start-replication).

Reviewer: ```suggestion
`source_cluster_uri` | The URI of the primary (source) cluster. The standby cluster connects to the primary cluster using this URI when [starting a replication stream]({% link {{ page.version.version }}/set-up-physical-cluster-replication.md %}#step-4-start-replication).
```

`replication_job_id` | The ID of the replication job.

> `replicated_time` | The latest timestamp at which the standby cluser has consistent data — that is, the latest time you can cut over to. This time advances while the replication proceeds without error. `replicated_time` is updated periodically (every `30s`).

Reviewer: ```suggestion
`replicated_time` | The latest timestamp at which the standby cluser has consistent data — that is, the latest time you can cut over to. This time advances automatically as long as the replication proceeds without error. `replicated_time` is updated periodically (every `30s`).
```

Author: Changed

`retained_time` | The earliest timestamp at which the standby cluster has consistent data — that is, the earliest time you can cut over to.

> `cutover_time` | The time at which the cutover will begin.

Reviewer: Is it worth mentioning that it. can be in the past or in the future?

Author: Done

`capability_name` | The [capability]({% link {{ page.version.version }}/create-virtual-cluster.md %}#capabilities) name.
`capability_value` | Whether the [capability]({% link {{ page.version.version }}/create-virtual-cluster.md %}#capabilities) is enabled for the virtual cluster.

{% comment %}Add responses to include for SQL statement page{% endcomment %}

### Data state

State      | Description
-----------+----------------
`initializing replication` | The replication job is completing the initial scan of data from the primary cluster before it starts replicating data in real time.
`ready` | The virtual cluster's data is ready for use.
`replicating` | The replication job has started and is replicating data.

> `replication paused` | The replication job is paused due to an error or a manual request with `ALTER VIRTUAL CLUSTER ... PAUSE REPLICATION`.

Reviewer: Would it be obvious whether it was due to an error or a manual pause?

> `replication pending cutover` | The replication job is running and a cutover time has been set. Once the the replication reaches the cutover time, the will begin automatically.

Reviewer: ```suggestion
`replication pending cutover` | The replication job is running and the cutover time has been set to a time in the future. Once the the replication reaches the cutover time, the will begin automatically.
```

Author: Changed

`replication cutting over` | The job has started cutting over. The cutover time can no longer be changed. Once complete, the virtual cluster will be available for use with `ALTER VIRTUAL CLUSTER ... START SHARED SERVICE`.
`replication error` | An error has occurred. You can find more detail in the error message and the logs.

## Examples

### Show all virtual clusters
{% comment %}this code block and output could be in an include. This is in several place e.g., cutover page, monitoring{% endcomment %}

List all of the virtual clusters:

{% include_cached copy-clipboard.html %}
~~~ sql
SHOW VIRTUAL CLUSTERS;
~~~

### Show a virtual cluster

{% comment %}this code block and output could be in an include. This is in several place e.g., cutover page{% endcomment %}

> To list a particular virtual cluster:

Reviewer: ```suggestion
To show more details about a particular virtual cluster:
```

Author: Done!
